src/track/beats.cpp

```diff
@@ -1,21 +1,156 @@
 
 #include "track/beats.h"
 
+#include "track/beatutils.h"
+
 namespace mixxx {
 
-int Beats::numBeatsInRange(double dStartSample, double dEndSample) {
+namespace {
+const int kFrameSize = 2;
+
+bool BeatLessThan(const track::io::Beat& beat1, const track::io::Beat& beat2) {
+    return beat1.frame_position() < beat2.frame_position();
+}
+
+double framesToSamples(const int frames) {
+    return frames * kFrameSize;
+}
+
+} // namespace
+
+Beats::Beats(const Track* track, const QVector<double>& beats, SINT iSampleRate)
+        : Beats(track, iSampleRate) {
+    if (beats.size() > 0) {
+        createFromBeatVector(beats);
+    }
+}
+
+Beats::Beats(const Track* track, const QByteArray& byteArray, SINT iSampleRate)
+        : Beats(track, iSampleRate) {
+    readByteArray(byteArray);
+}
+
+Beats::Beats(const Track* track, SINT iSampleRate)
+        : m_mutex(QMutex::Recursive),
+          m_track(track),
+          m_iSampleRate(iSampleRate == 0 ? m_track->getSampleRate() : iSampleRate),
+          m_dCachedBpm(0),
+          m_dLastFrame(0) {
+    // TODO(JVC) iSampleRate == 0 creates problems. Apparently only in tests
+
+    // BeatMap should live in the same thread as the track it is associated
+    // with.
+    moveToThread(track->thread());
+}
+
+// TODO(JVC) Do we really need a copy constructor??
+Beats::Beats(const Beats& other)
+        : m_mutex(QMutex::Recursive),
+          m_track(other.m_track),
+          m_subVersion(other.m_subVersion),
+          m_iSampleRate(other.m_iSampleRate),
+          m_dCachedBpm(other.m_dCachedBpm),
+          m_dLastFrame(other.m_dLastFrame),
+          m_beats(other.m_beats) {
+    moveToThread(m_track->thread());
+}
+
+void Beats::createFromBeatVector(const QVector<double>& beats) {
+    if (beats.isEmpty()) {
+        return;
+    }
+    double previous_beatpos = -1;
+    track::io::Beat beat;
+
+    foreach (double beatpos, beats) {
```

Comment:
`foreach` is a legacy Qt macro from before C++11. Use standard `for (double beatpos : beats)`